pkg/logs/log.go

```diff
@@ -86,11 +89,27 @@ type newLoggerFunc func() Logger
 // Logger defines the behavior of a log provider.
 type Logger interface {
 	Init(config string) error
-	WriteMsg(when time.Time, msg string, level int) error
+	WriteMsg(lm *LogMsg, opts ...common.SimpleKV) error
```

Comment:
I find that you use this to pass the `formatterFunc`, it looks a little bit strange. I think the `formatterFunc` should a new field of every implementation. And when we create a `Logger`, we set the value, like:
```go
type consoleWriter struct {
	OldLoggerAdapter
	lg       *logWriter
	fmtter LogFormatter
	Level    int  `json:"level"`
	Colorful bool `json:"color"` //this filed is useful only when system's terminal supports color
}
```

Reply:
And I think defining a new interface `LogFormatter ` is better than using `func(*LogMsg) string`.

Reply:
Yes I wasn't too sure if I should keep in opts but I wasn't entirely sure so I left it in.

And sure, using the `common.SimpleKV` is a bit weird but I was working off the draft PR you gave that used it. Your example above would be better

Reply:
In this way, we may need to define `BaseLogger` because I think many `Logger` may need `fmtter` field.
```go
type BaseLogger struct {
    fmtter...
}
```
And I am not sure whether we need to add `GetFmtter` and `SetFmtter` to `Logger` interface.
And I find that we could use template-method patten implement `WriteMsg` method, it may look like:
```go
func (l *BaseLogger) WriteMsg(m *LogMsg) {
 var  msgStr  string = l.fmtter.Format(m)
 l.writeMsg(msgStr)
}
```
I am not sure whether this is a good idea.